rust/sbd-server/src/lib.rs

```diff
@@ -115,40 +119,95 @@ impl Drop for SbdServer {
     }
 }
 
-async fn check_accept_connection(
-    _connect_permit: tokio::sync::OwnedSemaphorePermit,
-    config: Arc<Config>,
-    tls_config: Option<Arc<maybe_tls::TlsConfig>>,
-    ip_rate: Arc<ip_rate::IpRate>,
-    tcp: tokio::net::TcpStream,
+/// Convert an IP address to an IPv6 address.
+pub fn to_canonical_ip(ip: IpAddr) -> Arc<Ipv6Addr> {
+    Arc::new(match ip {
+        IpAddr::V4(ip) => ip.to_ipv6_mapped(),
+        IpAddr::V6(ip) => ip,
+    })
+}
+
+/// If the check passes, the canonical IP is returned, otherwise None and the connection should be
+/// dropped.
+pub async fn preflight_ip_check(
+    config: &Config,
+    ip_rate: &IpRate,
     addr: std::net::SocketAddr,
-    weak_cslot: cslot::WeakCSlot,
-) {
-    let raw_ip = Arc::new(match addr.ip() {
-        std::net::IpAddr::V4(ip) => ip.to_ipv6_mapped(),
-        std::net::IpAddr::V6(ip) => ip,
-    });
+) -> Option<Arc<Ipv6Addr>> {
+    let raw_ip = to_canonical_ip(addr.ip());
 
-    let mut calc_ip = raw_ip.clone();
+    let use_trusted_ip = config.trusted_ip_header.is_some();
+
+    if !use_trusted_ip {
+        // Do this check BEFORE handshake to avoid extra
+        // server process when capable.
+        // If we *are* behind a reverse proxy, we assume
+        // some amount of DDoS mitigation is happening there
+        // and thus we can accept a little more process overhead
+        if ip_rate.is_blocked(&raw_ip).await {
+            return None;
+        }
 
+        // Also precheck our rate limit, using up one byte
+        if !ip_rate.is_ok(&raw_ip, 1).await {
+            return None;
+        }
+    }
+
+    Some(raw_ip)
+}
+
+/// Handle an upgraded websocket connection.
+pub async fn handle_upgraded(
+    config: Arc<Config>,
+    ip_rate: Arc<IpRate>,
+    weak_cslot: WeakCSlot,
+    ws: Arc<impl SbdWebsocket>,
+    pub_key: PubKey,
+    calc_ip: Arc<Ipv6Addr>,
+) {
     let use_trusted_ip = config.trusted_ip_header.is_some();
 
-    let _ = tokio::time::timeout(config.idle_dur(), async {
-        if !use_trusted_ip {
-            // Do this check BEFORE handshake to avoid extra
-            // server process when capable.
-            // If we *are* behind a reverse proxy, we assume
-            // some amount of DDoS mitigation is happening there
-            // and thus we can accept a little more process overhead
-            if ip_rate.is_blocked(&raw_ip).await {
-                return;
-            }
+    // illegal pub key
+    if &pub_key.0[..28] == cmd::CMD_PREFIX {
```

Comment:
Is this the only format in which an illegal pub key can present?

Reply:
I think this is to prevent public keys overlapping with commands and confusing the server, I assume that any other invalid public key won't be able to decrypt messages so it's a bit pointless. I don't think we can tell anything about that though, without the private key or a message to test against, it's just a random byte array